### PR TITLE
[Commercial Server] (Base) PipelineLogger + PrometheusPipelineLogger

### PR DESCRIPTION
**Pull request by dbogunowicz**
This PR introduces the pipeline loggers (located in `src/deepsparse/loggers`).
In total, there are three new objects created:
- `BaseLogger` - an abstract class meant to define interfaces for the loggers that support various monitoring services APIs.
- `PrometheusLogger` - a logger that uses the Prometheus client to monitor the inference pipeline
- `ManagerLogger` - a logger manager that oversees all the pipeline loggers

Testing is covered by unit tests

Example use of Prometheus logger.

```python
pipeline_yolo = ...
logger = PrometheusLogger( port = 8000)
while True:
    input_image = ... # from external source
    out, input_schema, engine_input, inference_timing = pipeline_yolo.run_with_monitoring(images=input_image)
    logger.log_latency(pipeline_yolo.name, inference_timing)
    logger.log_data(pipeline_yolo.name, ...) # interface for logging data to be defined
```

Example use of Manager logger:

```python
loggers = [PrometheusLogger(), CloudWatchLogger(), ...]
logger_manager = ManagerLogger(loggers)
pipeline_yolo = ...
while True:
    input_image = ... # from external source
    out, input_schema, engine_input, inference_timing = pipeline_yolo.run_with_monitoring(images=input_image)
    logger.log_latency(pipeline_yolo.name, inference_timing)
    logger.log_data(pipeline_yolo.name, ...) # interface for logging data to be defined
```

### CODE REVIEW

File: src/deepsparse/pipeline_loggers/prometheus_pipeline_logger.py
Comment (bfineran) [inline src/deepsparse/pipeline_loggers/prometheus_pipeline_logger.py:71]: definitely don't want to log to console every time we log to prometheus, especially with no extra info

File: src/deepsparse/pipeline_logger.py
Comment (bfineran) [inline src/deepsparse/pipeline_logger.py:28]: need infrastructure to get a different logger instance for each pipeline

**Review by bfineran — CHANGES_REQUESTED** [review]
great first pass. seeing as how the implementations need to do some global setup, I think it doesn't make sense for us to initialize new loggers for each pipeline anymore. Instead, let's make `pipeline_name` a required argument of each of the `log_` methods (ie `log_latency("pipeline_1", ...)`)

this will also make the integration at the server level easier as we don't need to reconcile creation and management of logger objects for every combination of logger impl and pipeline entrypoint

let's aim for to support a UX for something like this:

```python
# create logger and pipelines
logger = LoggerManager(["prometheus", "elastic_search"])
pipelines = [Pipeline.from_config(config) for config in pipeline_configs]

def create_routes(pipelines, logger):
    # define route for each pipeline, pipelines share the global logger
    for pipeline in pipelines:
        @get(ROUTE_NAME)
         def pipeline_route(inputs):
                # run pipeline
                results, timings, data = pipeline.run_with_monitoring(inputs)  # pathway name here tbd
                # logger manager handles logging to each integration by pipeline name
                logger.log_timings(pipeline.name, timings)
                logger.log_data(pipeline.name, data)
                return results
```

File: src/deepsparse/pipeline_loggers/prometheus_pipeline_logger.py
Comment (dbogunowicz) [inline src/deepsparse/pipeline_loggers/prometheus_pipeline_logger.py:86]: private method, move down

File: src/deepsparse/pipeline_logger.py
Comment (corey-nm) [inline src/deepsparse/pipeline_logger.py:34]: Related to my other comment.
Let's make this an abstract property for subclasses to implement, then it doesn't have to given to this class' constructor. And then the prometheus one can just define it as a property instead of having a global variable for it.

File: src/deepsparse/pipeline_logger.py
Comment (corey-nm) [inline src/deepsparse/pipeline_logger.py:38]: Let's move this into the subclasses, esp since this is an abstract class, no one will have an instance of just this

File: src/deepsparse/pipeline_logger.py
Comment (corey-nm) [inline src/deepsparse/pipeline_logger.py:55]: What's the expected order of calls for log_data and log_latency? Are they both required to be called or is log_data optional?

File: src/deepsparse/pipeline_loggers/manager.py
Comment (corey-nm) [inline src/deepsparse/pipeline_loggers/manager.py:71]: Are we expecting people to pass in their own set of supported loggers? I think this should not be a parameter and should just be a instance variable.

Additionally if you make the identifier a abstract class property you could do:
```python
loggers = [PrometheusLogger]
self._supported_loggers: Dict[str, Type[PipelineLogger]] = {logger.identifier: logger}
```

File: src/deepsparse/pipeline_loggers/manager.py
Comment (corey-nm) [inline src/deepsparse/pipeline_loggers/manager.py:72]: In tandem with the above comment you could remove _setup_loggers for this:
```python
# check loggers
for identifier in logger_identifiers:
    if identifier not in self._supported_loggers:
                raise ValueError(
                    "Attempting to create a pipeline logger with an "
                    f"unknown identifier: {identifier}. Supported "
                    f"identifiers are: {self._supported_loggers}"
                )

self._loggers = list(map(self._supported_loggers.get, logger_identifiers))
```

File: src/deepsparse/pipeline_loggers/manager.py
Comment (corey-nm) [inline src/deepsparse/pipeline_loggers/manager.py:68]: I'm going to recommend a larger change that is to just pass in the list of PipelineLoggers directly here. That will do two things:
1. You can probably just remove the identifier all together in that class (which would remove a big portion of this)
2. Make it more clear that LoggerManager is a "meta" PipelineLogger that is built up of a list of sub pipeline loggers.

File: src/deepsparse/pipeline_loggers/manager.py
Comment (corey-nm) [inline src/deepsparse/pipeline_loggers/manager.py:34]: Suggestions:
1. Subclass from PipelineLogger
2. Rename to something like `MultiLogger` or `SequentialLogger` or something that captures that this is a pipeline logger built up of a list of other pipeline loggers

**Review by corey-nm — CHANGES_REQUESTED** [review]
I think this needs a bit more work structurally. Seems like it will be straightforward to use prometheus though!

File: src/deepsparse/pipeline_loggers/manager.py
Comment (dbogunowicz) [reply src/deepsparse/pipeline_loggers/manager.py:68]: I like the idea, but I wonder how this would work, given that we use human-readable (writable) config to create the loggers. What I am saying is, that it would be natural to pass identifiers (string), directly parsed from the config, rather then logger classes, to create the manager.

File: src/deepsparse/pipeline_loggers/manager.py
Comment (corey-nm) [reply src/deepsparse/pipeline_loggers/manager.py:34]: Mainly to make it clear to readers & others that that is what it is. I see it the same as if PrometheusLogger didn't subclass PipelineLogger

File: src/deepsparse/pipeline_loggers/manager.py
Comment (corey-nm) [reply src/deepsparse/pipeline_loggers/manager.py:68]: I think in that case the server should do the mapping from identifier or name to the class and construct the classes. This keeps the human readable stuff isolated into server, so it doesn't infect this code.

As far as how the server does the mapping it could either be strings like this, or we could even have users use the class name and then do a `getattr(deepsparse.pipeline_loggers, name)`, where name could be `PrometheusLogger`.

File: src/deepsparse/pipeline_loggers/manager.py
Comment (dbogunowicz) [reply src/deepsparse/pipeline_loggers/manager.py:71]: In general, your whole set of bits of advice helps tremendously. However, the code above would not work. Properties are "properties" (duh!) of instances and not classes. There is a hack to make them global though, will try that since I really like the simplifications you are introducing

File: src/deepsparse/pipeline_loggers/manager.py
Comment (dbogunowicz) [reply src/deepsparse/pipeline_loggers/manager.py:34]: Ok, this makes sense.

File: src/deepsparse/pipeline_logger.py
Comment (dbogunowicz) [reply src/deepsparse/pipeline_logger.py:1]: Keeping it one level above, so that it will be harder to run into circular dependency import issue when creating new loggers

File: src/deepsparse/pipeline_loggers/multi_logger.py
Comment (dbogunowicz) [inline src/deepsparse/pipeline_loggers/multi_logger.py:37]:
```suggestion
    The envisioned lifecycle of a multilogger
```

File: src/deepsparse/pipeline_loggers/multi_logger.py
Comment (dbogunowicz) [inline src/deepsparse/pipeline_loggers/multi_logger.py:58]:
```suggestion
        will be created within the scope of the logger. This can be either a
```

File: tests/deepsparse/pipeline_loggers/test_multilogger.py
Comment (dbogunowicz) [inline tests/deepsparse/pipeline_loggers/test_multilogger.py:18]: There are some naming problems that I will fix after the PR is approved.

File: src/deepsparse/pipeline_loggers/multi_logger.py
Comment (dbogunowicz) [inline src/deepsparse/pipeline_loggers/multi_logger.py:38]: There are some naming problems that I will fix after the PR is approved.

File: src/deepsparse/pipeline_loggers/multi_logger.py
Comment (dbogunowicz) [inline src/deepsparse/pipeline_loggers/multi_logger.py:105]:
```suggestion
    ) -> Dict[str, Type[PipelineLogger]]:
```

File: src/deepsparse/pipeline_logger.py
Comment (corey-nm) [inline src/deepsparse/pipeline_logger.py:38]: Maybe we can just use a method instead of property?
```suggestion
    @staticmethod
    @abstractmethod
    def identifier(self) -> str:
```

File: src/deepsparse/pipeline_loggers/multi_logger.py
Comment (corey-nm) [inline src/deepsparse/pipeline_loggers/multi_logger.py:65]: We could just pass in the list of instantiated logger objects directly into this class `loggers: List[PipelineLogger]`, and move the creation/parsing logic (`self._setup`, `self._start`) outside of this class to somewhere else (the server I guess?)

Then this class would be a really tightly scoped util class to easily wrap multiple loggers. It'd be easier to test to since you won't have to go through strings and could just pass in the loggers you want to use.

Comment (dbogunowicz) [reply src/deepsparse/pipeline_loggers/multi_logger.py:65]: I like this idea a lot!

File: src/deepsparse/pipeline_logger.py
Comment (bfineran) [reply src/deepsparse/pipeline_logger.py:1]: I don't think we'd run into dependency issues as long as we use the full module path.

let's also rename the `pipeline_loggers` directory to `logging` or `loggers`

File: src/deepsparse/pipeline_logger.py
Comment (bfineran) [inline src/deepsparse/pipeline_logger.py:22]: nit: docstring

File: src/deepsparse/pipeline_logger.py
Comment (bfineran) [inline src/deepsparse/pipeline_logger.py:57]:
```suggestion
        """
        raise NotImplementedError()
```

File: src/deepsparse/pipeline_loggers/multi_logger.py
Comment (bfineran) [reply src/deepsparse/pipeline_loggers/multi_logger.py:38]: sounds good, not a fan of `MultiLogger` name

Comment (bfineran) [inline src/deepsparse/pipeline_loggers/multi_logger.py:38]: nit: pretty short line breaks here

File: src/deepsparse/pipeline_loggers/multi_logger.py
Comment (bfineran) [inline src/deepsparse/pipeline_loggers/multi_logger.py:104]: need to think if we can do something more extensible here than string names to the desired loggers - ie if there's stuff we need to configure with an integration like prometheus's port, save directory, etc - we'd like to be able to control that at the server (and even user) level. 

so just taking a list of constructed logger objects here and ignoring identifiers might be the way to go, leaving the onus of construction to the server

File: src/deepsparse/pipeline_loggers/prometheus_pipeline_logger.py
Comment (bfineran) [inline src/deepsparse/pipeline_loggers/prometheus_pipeline_logger.py:151]: can use a default dict (of dict) here

**Review by bfineran — CHANGES_REQUESTED** [review]

File: src/deepsparse/pipeline_logger.py
Comment (dbogunowicz) [reply src/deepsparse/pipeline_logger.py:57]: Corey pointed out: "For abstract methods, you can actually just have the docstring and it'll auto do the not implemented thing."
Tbh, I could not find the confirmation of this statement on the internet, but it's working fine, so I guess this may be true.

File: src/deepsparse/pipeline_loggers/multi_logger.py
Comment (dbogunowicz) [reply src/deepsparse/pipeline_loggers/multi_logger.py:104]: good point!

File: src/deepsparse/pipeline_logger.py
Comment (bfineran) [reply src/deepsparse/pipeline_logger.py]: would prefer to make this explicit

**Review by bfineran — COMMENTED** [review]
lgtm pending comment and PR description

**Review by bfineran — APPROVED** [review]
LGTM, please add examples of prometheus usage/access in the description

File: src/deepsparse/loggers/prometheus_logger.py
Comment (corey-nm) [inline src/deepsparse/loggers/prometheus_logger.py:87]: Is `self._counter` only used here to get the number of times logger has been called? Can self._counter just be a single value instead of a dictionary?

**Review by corey-nm — APPROVED** [review]
LGTM 🎉

File: src/deepsparse/loggers/prometheus_logger.py
Comment (dbogunowicz) [reply src/deepsparse/loggers/prometheus_logger.py:87]: self._counter counts how many inferences were done per pipeline
self.counter sums them up and returns the total count of interences of the logger

Comment (corey-nm) [reply src/deepsparse/loggers/prometheus_logger.py:87]: It looks like self._counter[pipeline_name] is incremented once per log_latency call - can you just have an integer attribute that you increment instead of the dictionary? Then you could just use `self._counter` directly instead of doing the sum of values

**Pull request merged**
